### ISSUE TITLE
Latex Parser interpreting 'e' and 'pi' as free symbols rather than special characters

### ISSUE THREAD

**Issue by Hendrick-deHaan** [sympy/sympy#20306]
Using the parse_latex command, 'e' and 'pi' are not interpreted correctly. A work around of 

result = result.subs({"e":sy.E, "pi":sy.pi})

resolves the issue as suggested here: 

https://groups.google.com/g/sympy/c/vw6PnI5O4n0

**Comment by sylee957** [comment]
I don't think that e and pi have to be implemented to the parser directly, but as a post-processing mechanic. If we start to support e and pi directly to parsers, there can be other problems of scripted symbols like `\pi_1` which we can't necessarily comprehend as constants.

**Comment by jaded0** [comment]
Well, that's just the fundamental problem we run into when we try to convert latex to sympy. Lots of assumptions have to be made.
The most common interpretation of e and pi symbols is the expected behavior when parsing latex, as demonstrated by the edits made to the projects latex2sympy and latex2sympy2. 
I'm not familiar with the usage of `\pi_1`, though, or the parsers. In my opinion, a default setting that allows you to specify symbols to not convert to constants would be best.

**Comment by jaded0** [comment]
Or, perhaps, a separate function that replaces common symbols for their constants.

**Comment by wermos** [comment]
What was the final verdict for this issue? The reason I am asking is because I want to solve this issue for the Lark-based LaTeX parser.

> I don't think that e and pi have to be implemented to the parser directly, but as a post-processing mechanic. If we start to support e and pi directly to parsers, there can be other problems of scripted symbols like `\pi_1` which we can't necessarily comprehend as constants.

This is a valid concern. However, because of the way I have implemented the Lark-based LaTeX parser, something like `e_1` would be a `BASIC_SUBSCRIPTED_SYMBOL`, and something like `\pi_1` would be a `GREEK_SUBSCRIPTED_SYMBOL`.

What this means is that I can differentiate between a naked `\pi` or `e` and a `\pi` or `e` with a subscript. This means I can add a special code path in `SYMBOL` (which captures a naked `e`) and `GREEK_SYMBOL` (which captures a naked `\pi`) to convert it into `sympy.E` and `sympy.pi`.

cc @Upabjojr and @asmeurer for thoughts.

**Comment by asmeurer** [comment]
e and \pi should obviously default to returning the numerical constants, at least when they are in a context where that seems like what they should be. As to whether it should be done in the parser itself or as a post-processor, I don't know how the parser is structured, but it would be nice for people to be able to override that. If the parser allows overriding that at the grammar level, then I guess it doesn't matter. 

Just note that it wouldn't really work to override it after the fact because e or pi might have been evaluated in some way. Or for instance, \pi might really be the function `primepi()`.

**Comment by sylee957** [comment]
My opinion is that we can take 2 most general approach

- First, everything should initially give `Symbol` everywhere and and we can let users substitute the `Symbol` with custom constant dictionary
  (In the reverse way, if the default of `pi` is constant than Symbol, `sin(pi)` could already be evaluated to number and users can't recover the free symbol as `pi`)
- Second, the transformer can be inherited by the users and they can just override some symbols there
```
class MyTransformer(...):
    def e(self, ...):
        return ...

    def pi(self, ...):
        return ...
```

I think that these two options are equally general,

**Comment by asmeurer** [comment]
You need to be able to modify the grammar. Something like `\pi(n + 1)` would parse as `pi*(n + 1)` but if `\pi` is supposed to be a function `primepi` it should parse as `primepi(n + 1)`. That can't be fixed after the parsing has already happened with a multiplication.

**Comment by sylee957** [comment]
One idea I have is that we can just parse everything as ‘function’ initially and selectively apply things as multiplication later

**Comment by asmeurer** [comment]
This sort of thing could come up in more complex ways than just multiplication vs. function application. LaTeX represents handwritten mathematics, which is very non-rigorous and has lots of abuses of notation and shortcuts that differ across domains. As another example, sometimes a superscript is a power, sometimes it is an index, sometimes it is just part of the variable name.

The best we could hope to do is to parse everything into some intermediate latex-AST, which has sufficient information to round-trip back to the LaTeX (modulo unimportant details like whitespace), and then define a way to transform that AST into SymPy in a user-extensible way.

**Comment by wermos** [comment]
> As to whether it should be done in the parser itself or as a post-processor

I'm not sure what such a post-processor would look like. Can you guys elaborate on that and tell me what you had in mind?

> Second, the transformer can be inherited by the users and they can just override some symbols there
> ```python
> class MyTransformer(...):
>    def e(self, ...):
>         return ...
>
>     def pi(self, ...):
>         return ...
> ```

I'd like to point out that this method would not work in practice. The transformer can only process rules and nodes. `e` and `\pi` don't have their own nodes. `e` is part of the `SYMBOL` group, and `\pi` is part of the `GREEK_SYMBOL` group. What this means is, in order to make the changes that I was suggesting, I would modify the transformer functions for these and add a special case.

On the flip side, this means that in order to make the transformer not return `sympy.E` for `e` and `sympy.pi` for `\pi`, they would have to modify the original transformer. Inheriting from the transformer class and defining two functions called `e` and `pi` wouldn't help. They would either have to modify the code for `GREEK_SYMBOL` and `SYMBOL`, or override those methods in the child class.

> One idea I have is that we can just parse everything as ‘function’ initially and selectively apply things as multiplication later

I didn't quite understand this idea. Could you elaborate on this please?

**Comment by sylee957** [comment]
> e is part of the SYMBOL group, and \pi is part of the GREEK_SYMBOL group.

I think that that only depends on your implementation.
It is possible to define the rules of alphabets individually

a: “a”
…
z: “z”

**Comment by wermos** [comment]
> It is possible to define the rules of alphabets individually

Yes, but code duplication would increase greatly. Also, we'd lose most of our existing functionality: Something like `sin x` would bind to both `s*i*n*x` and the trigonometric function. In short, I don't think this is a viable way to go about this.

I already have an issue with `\negthinspace` and friends because the `\ne` binds to the `NOT_EQUAL` relational operator, so it thinks that everything that comes after the `\ne` is implicit multiplication.

**Comment by sylee957** [comment]
For specific case, ‘sin x’, I don’t think that it have problem because backslash disambiguates $\sin x$ and $sin x$, in fact, $\sin x$ is wrong

The problem is doesn’t go away because you anyway put the branching logic depends on the alphabets in transformer, and I favor to write the if-else logic as much into grammar than inside transformer, because parsing grammar has better decidability capability than writing in general programming languages.

you can write the regex like ‘a-df-z’ if you want to only trivialize ‘e’ from there.

**Comment by sylee957** [comment]
About how to do the function to switch over multiplication,
you may play with overriding function_applied
with 
Symbol(tokens[0].name) * list_of_expressions[0] 
depending on name

**Comment by Upabjojr** [comment]
> You need to be able to modify the grammar. Something like `\pi(n + 1)` would parse as `pi*(n + 1)` but if `\pi` is supposed to be a function `primepi` it should parse as `primepi(n + 1)`. That can't be fixed after the parsing has already happened with a multiplication.

Maybe `a(expr)` should have its own node different from both multiplication and function application, so that disambiguation may happen at the transformer stage.

**Comment by wermos** [comment]
What is `a(expr)` here?

**Comment by asmeurer** [comment]
> Maybe a(expr) should have its own node different from both multiplication and function application, so that disambiguation may happen at the transformer stage.

Again, this issue is much more general than just multiplication vs. function application. It could come up anywhere in the grammar. Your idea is basically the same as my suggestion:

> The best we could hope to do is to parse everything into some intermediate latex-AST, which has sufficient information to round-trip back to the LaTeX (modulo unimportant details like whitespace), and then define a way to transform that AST into SymPy in a user-extensible way.

A LaTeX AST (or CST) would be useful for other purposes too. But this would be complicated to do and it's not the only way to solve the problem.

**Comment by sylee957** [comment]
What about writing different lark files with custom functions turned off, or applied with restricted set of symbols (for example, fgh)
There can be some gain of trivializing some part of grammar.

For example, I had requested users to only use fgh as valid fuction names, or xyz as symbol names, and this made the implementation more easier, because the grammar never had the conflicts.

**Comment by wermos** [comment]
> What about writing different lark files with custom functions turned off

Can you explain what you mean by "custom functions" in this context?

>applied with restricted set of symbols

So, are you saying that we ask users to use only f, g, and h for function application and things like x, y, and z for variable names?

> For example, I had requested users to only use fgh as valid fuction names

Where did you make this request?

**Comment by sylee957** [comment]
> A LaTeX AST (or CST) would be useful for other purposes too.

Specially for lark, `.parse` without transformer is free lunch for AST.
I just believe that good parser individually, is just reusable in that regards.

**Comment by sylee957** [comment]
> So, are you saying that we ask users to use only f, g, and h for function application and things like x, y, and z for variable names?

I would mean what about allowing users to use `%override` or `%extend` syntax for things like `SYMBOL`. They can define or restrict the names of symbols if they don't need generality.

**Comment by Upabjojr** [comment]
> What about writing different lark files with custom functions turned off, or applied with restricted set of symbols (for example, fgh) There can be some gain of trivializing some part of grammar.
> 
> For example, I had requested users to only use fgh as valid fuction names, or xyz as symbol names, and this made the implementation more easier, because the grammar never had the conflicts.

What about having a template for the lark grammar instead of different lark files? The rules may then be applied by properly filling the template and getting a customized lark file.

**Comment by wermos** [comment]
> > A LaTeX AST (or CST) would be useful for other purposes too.
>
> Specially for lark, .parse without transformer is free lunch for AST.
I just believe that good parser individually, is just reusable in that regards.

What @sylee957 said is correct. Until we use the `.parse` function, which applies the Tranformer class on the AST, we have actually, in essence, made a LaTeX AST.

> Again, this issue is much more general than just multiplication vs. function application. It could come up anywhere in the grammar. Your idea is basically the same as my suggestion:
>
> > The best we could hope to do is to parse everything into some intermediate latex-AST, which has sufficient information to round-trip back to the LaTeX (modulo unimportant details like whitespace), and then define a way to transform that AST into SymPy in a user-extensible way.

So, I guess the question boils down to two things:
1. What exactly do we want the LaTeX parser to do? I.e., what parts do we all agree should be part of the LaTeX parser, perhaps with some options available for customizing its behavior, and what parts should be left to the user?
2. How do we want to support user customization? How much flexibility do we want to allow?

**Comment by sylee957** [comment]
- At the point of defining function symbols, I think that we can only parse f, g, h as functions, could let users to override function name tokens in grammar if they need other functions. Because I think that it is technically difficult to use _ambig properly.

- For pi, e, i symbols, we can add boolean keywords to let them parse to constant and symbols. We just let usual mathematical convention as default.
(We may let users to use tau as pi, or j as imaginary constant, however, we don’t need to add everything as options because they can be done in other ways, like subs(tau, 2*pi))

**Comment by wermos** [comment]
That sounds like a good solution. Another added bonus is that it is easy to implement and has clearly defined boundaries for what we will do by default, and what the user needs to do to achieve anything extra they want to be done.

**Comment by sylee957** [comment]
And I think that LaTeX CST (That is generated by `parse`) can be another topic about to make good use.

I even don't think that SymPy AST is enough,
for instance, if I have extra parenthsis like $(x)$, SymPy will likely not be able to represent extra parenthesis like that.
However, LaTeX CST can retain such information, so I wonder if this can be the direction to use LaTeX CST.

**Comment by wermos** [comment]
> However, LaTeX CST can retain such information, so I wonder if this can be the direction to use LaTeX CST.

This doesn't sound like a good idea. LaTeX is a typesetting language so it needs to keep track of anything and everything that would have a bearing on the final visual output. As such, absent any other external context, extra parentheses surrounding a variable (or `Symbol`) bears no mathematical meaning. That's why a CAS doesn't need to keep track of such fluff.

**Comment by wermos** [comment]
> Again, this issue is much more general than just multiplication vs. function application. It could come up anywhere in the grammar.

@asmeurer You mentioned that this issue could crop up anywhere in the grammar. Could you elaborate on that, and perhaps give us some examples?

**Comment by sylee957** [comment]
> That's why a CAS doesn't need to keep track of such fluff.

I had similar opinion like this, but I think that that is specific to SymPy. 
mathjs, for instance, have ParenthesisNode, and they may have solved the issues in the way we overlooked

https://mathjs.org/docs/expressions/expression_trees.html#parenthesisnode

**Comment by jackdeye** [comment]
Hi, I was wondering if this ambiguity was ever resolved? I noticed that the parser fails most test cases in `test_applied_function_expression` and saw that it was because it was interpreting `f(x)` as SYMBOL * SYMBOL. 

In #19527, I saw that it was suggested that the user be able to specify behavior with this ambiguity, and I agree. Would the following functionality fix this design issue?
```python
class TransformToSymPyExpr(Transformer):
def __init__(
        self,
        known_functions: Optional[Set[str]] = None,
        known_symbols: Optional[Set[str]] = None,
        use_spacing_heuristic: bool = True,
        ambiguity_resolution: Literal["function", "multiplication", "error"] = "function",
        visit_tokens: bool = True
    ):
...
```
Changes to other files might be necessary; however, I am envisioning the following functionality:
```python
        >>> parser = ConfigurableLarkLaTeXParser(
        ...     known_functions={"f", "g", "h"},
        ...     known_symbols={"a", "b", "c"},
        ...     ambiguity_resolution="error"
        ... )
        >>> expr = parser.parse("f(x)")  # -> Function('f')(x)
        >>> expr = parser.parse("a(x)")  # -> a * x  
        >>> expr = parser.parse("unknown(x)")  # -> raises error
```
This would allow users to explicitly specify which names should be treated as functions vs symbols as well as use spacing heuristics (a (x) vs a(x)) to disambiguate.

**Comment by wermos** [comment]
There has been no work on this issue for some time.

The proposed solution looks reasonable to me and looks like a reasonable default that most users wouldn't have any issues with. Ultimately, parsing LaTeX is a difficult task so producing a solution that accounts for 100% of the pathological use-cases might not be possible. With that in mind, these solutions which work in the way that most people would accept are acceptable to me.

What exactly does `use_spacing_heuristic` mean and do? Also, (if I understood it correctly), it looks like this option might have something to do with [this failing test](https://github.com/sympy/sympy/blob/97e513139dee148567782c4b15fb56fa61e3c1ac/sympy/parsing/tests/test_latex_lark.py#L837-L842), so if we can get that test to pass, that would be great.

I invite @sylee957 and @asmeurer as previous participants in this discussion to share their thoughts and opinions.

**Comment by jackdeye** [comment]
Thank you for the insight. My intent for `use_spacing_heuristic` was how `f (x)` vs `f(x)` should be interpreted (or any white space between f and open paren), where, with it enabled, it might interpret the `f (x)` as multiplication as opposed to a function (and vice versa). However, I don't think people tend to write `$f\left(x \right)$` frequently, as personally I would write `$f \left(x \right)` so the exact implementation might change.

I believe that the failing test case is because there is only one [grammar for multiplication](https://github.com/sympy/sympy/blob/97e513139dee148567782c4b15fb56fa61e3c1ac/sympy/parsing/latex/lark/grammar/latex.lark#L214), which does not allow for white space.

I will try to submit a pr for at least one of these issues sometime this week.

**Comment by wermos** [comment]
The current version of the Lark-based parser ignores all whitespace, so it "sees" `f (x)` and `f(x)` as the same thing.

> I will try to submit a pr for at least one of these issues sometime this week.

Perfect! When you open a PR, please tag me for review.
